shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs

```diff
@@ -1085,23 +1104,123 @@ updateNonMyopic nm rPot newLikelihoods =
         <> newPerf
     updatedLikelihoods = Map.mapWithKey performance newLikelihoods
 
--- | Create a reward update
-createRUpd ::
-  forall era.
+-- =========================================
+-- Useful data to pass around intermediate results
+
+-- | To pulse the reward update, we need a snap shot of the EpochState particular to this computation
+data RewardSnapShot crypto = RewardSnapShot
+  { rewSnapshots :: !(SnapShots crypto),
+    rewa0 :: Rational,
+    rewnOpt :: Natural,
+    rewprotocolVersion :: ProtVer,
+    rewNonMyopic :: !(NonMyopic crypto),
+    rewDeltaR1 :: !Coin, -- deltaR1
+    rewR :: !Coin, -- r
+    rewDeltaT1 :: !Coin, -- deltaT1
+    rewTotalStake :: !Coin, -- totalStake
+    rewRPot :: !Coin -- rPot
+  }
+  deriving (Show, Eq, Generic)
+
+-- instance HasField "_d" (RewardSnapShot crypto) UnitInterval where
+-- instance HasField "_tau" (RewardSnapShot crypto) UnitInterval where
+-- instance HasField "_rho" (RewardSnapShot crypto) UnitInterval where
+instance HasField "_a0" (RewardSnapShot crypto) Rational where
+  getField x = rewa0 x
+
+instance HasField "_nOpt" (RewardSnapShot crypto) Natural where
+  getField x = rewnOpt x
+
+instance HasField "_protocolVersion" (RewardSnapShot crypto) ProtVer where
+  getField x = rewprotocolVersion x
+
+instance NoThunks (RewardSnapShot crypto)
+
+instance NFData (RewardSnapShot crypto)
+
+instance CC.Crypto crypto => ToCBOR (RewardSnapShot crypto) where
+  toCBOR (RewardSnapShot ss a0 nopt ver nm dr1 r dt1 tot pot) =
+    encode
+      ( Rec RewardSnapShot !> To ss !> To a0 !> To nopt !> To ver !> To nm !> To dr1
+          !> To r
+          !> To dt1
+          !> To tot
+          !> To pot
+      )
+
+instance CC.Crypto crypto => FromCBOR (RewardSnapShot crypto) where
+  fromCBOR = decode (RecD RewardSnapShot <! From <! From <! From <! From <! From <! From <! From <! From <! From <! From)
+
+-- | State used in the STS rules
+data PulsingRewUpdate m crypto
```

Comment:
Given there's now quite some internal structure, could we move this to a separate module so `LedgerState` doesn't get even bigger

Reply:
erm, maybe?

Reply:
Could live in `Rewards`? Unless that hits a dep problem

Reply:
Most pulsing related stuff (FreeVars,   RewardSnapShot, RewardPulser, etc)
has been moved to   Shelley.Spec.Ledger.RewardUpdate
Everyhting is much cleaner.